File: python/haffman.py
from operator import itemgetter
from collections import deque

class Tree:
    def __init__(self, cargo, left=None, right=None):
        self.cargo = cargo
        self.left  = left
        self.right = right

    def __str__(self):
        return str(self.cargo)
        
    def __getitem__(self, key):
        return self.cargo[key]

def traverseTree(tree, cur_code, haff_dict):
    if cur_code != '' and "cargo_" not in tree[1]:
        haff_dict[tree[1]] = cur_code
    try:
        traverseTree(tree.left, cur_code + "0", haff_dict)
        traverseTree(tree.right, cur_code + "1", haff_dict)
    except:
        return 

def haff(input_str):
    result_str = ''
    
    queue_with_priority = deque()
    # get priority and into to deque
    for c in input_str:
        is_find = False
        for i in queue_with_priority:
            if c == i[-1]:
                i[0] += 1
                is_find = True
        if not is_find:
            queue_with_priority.append([1, c])
            
    # sort deque
    queue_with_priority = deque(sorted(queue_with_priority, key=itemgetter(0)))
    
    # get tree
    c_num = 1
    while len(queue_with_priority) != 1:
        i = queue_with_priority.popleft()
        j = queue_with_priority.popleft()
        t = Tree([i[0] + j[0], 'cargo_{0}'.format(c_num)], i, j)
        c_num += 1
        queue_with_priority.appendleft(t)
        queue_with_priority = deque(sorted(queue_with_priority, key=itemgetter(0)))
    
    # traverse a tree and create Haffman dict
    haff_dict = {}
    if type(queue_with_priority[0]) is list and len(queue_with_priority) == 1:
        haff_dict[queue_with_priority[0][1]] = "0"
    else:    
        traverseTree(queue_with_priority[0], "", haff_dict)

    # enc string
    for a in input_str:
       result_str += haff_dict[a]

    return result_str

test_str = 'accepted'
print(haff(test_str))

def decodeHaffman():
    # get data
    haff_dict = {}
    n, str_len = input().split()
    for i in range(0, int(n)):
        code = input().split(': ')
        haff_dict[code[-1]] = code[0]
        
    enc_str = str(input())
    
    # decode string
    result_str = ""
    cur_char = ""
    for a in enc_str:
        cur_char += a
        if cur_char in haff_dict.keys():
            result_str += haff_dict[cur_char]
            cur_char = ""
    
    print(result_str)

File: c++/haffman.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
#include <map>

void DecodeHaffman(){
    // get data
    int n, str_len;
    std::cin >> n >> str_len;
    std::map<std::string, std::string> haff_dict;;
    
    // create haffman dict
    for(int i = 0; i < n; i++){
        std::string value, key;
        std::cin >> value >> key;
        value = value[0];
        haff_dict.insert(std::pair<std::string, std::string>(key, value));
    }
    
    std::string encoded_string;
    std::cin>>encoded_string;
    
    // decode string
    std::string res_string = "";
    std::string cur_code = "";
    
    for(auto& c : encoded_string){
        cur_code += c;
        if(haff_dict.count(cur_code) != 0){
            res_string += haff_dict[cur_code];
            cur_code = "";
        }
    }
    
    std::cout << res_string << std::endl;
}

struct Node{
    std::string key;
    int priority;
    Node *left;
    Node *right;
};

// Traverse tree and fill haffman dict
void TraverseTree(Node* leaf, std::string cur_code, std::map<std::string, std::string> *haffman_dict){
    if(leaf == NULL) return;
    // cargo check
    if(leaf->key.find("cargo_") == std::string::npos){
        // null pointer check
        if(haffman_dict){
		haffman_dict->insert(std::pair<std::string, std::string>(leaf->key, cur_code));
	}
    }

    TraverseTree(leaf->left, cur_code + "0", haffman_dict);
    TraverseTree(leaf->right, cur_code + "1", haffman_dict);
}

// lambda func for sorting
bool SortDescSecond(const Node* a, const Node* b) { 
    return (a->priority > b->priority); 
} 

// create new Node
Node* CreateNode(std::string key, int priority, Node* left_leaf, Node* right_leaf){
    Node* new_node = new Node;
    new_node->key = key;
    new_node->priority = priority;
    new_node->left = left_leaf;
    new_node->right = right_leaf;

    return new_node;
}

std::string EncodeHaffman(std::string input_str) {
    std::string encoded_str = "";
    
    // get priority
    std::vector<Node*> queue_with_priority;
    
    for(auto& c : input_str){
        bool not_find = true;
        std::string s;
        s.push_back(c);
        
        for(auto& q : queue_with_priority){
            if(q->key == s){
                q->priority += 1;
                not_find = false;
                break;
            }
        }
        
        if(not_find){
            queue_with_priority.push_back(CreateNode(s, 1, NULL, NULL));
        }
    }
    
    // sort queue
    sort(queue_with_priority.begin(), queue_with_priority.end(), SortDescSecond);
    
    // build binary tree
    int cargo_count = 1;
    while(queue_with_priority.size() != 1){
	Node* left_leaf = queue_with_priority.back();
        queue_with_priority.pop_back();
   	Node* right_leaf = queue_with_priority.back();
        queue_with_priority.pop_back();

        std::string cargo_key = "cargo_" + std::to_string(cargo_count);
        int cargo_priority = left_leaf->priority + right_leaf->priority;

	queue_with_priority.push_back(CreateNode(cargo_key, cargo_priority, left_leaf, right_leaf));
	sort(queue_with_priority.begin(), queue_with_priority.end(), SortDescSecond);

        cargo_count++;
    }
    
    // generate haffman dict
    std::map<std::string, std::string> haffman_dict;
    TraverseTree(queue_with_priority.back(), "", &haffman_dict);
    
    // for one element
    if(queue_with_priority.size() == 1 && haffman_dict.size() == 1){
        haffman_dict[queue_with_priority.back()->key] = "0";
    }
    
    // encoding input string
    for(auto& c : input_str){
        std::string s;
        s.push_back(c);
	if(haffman_dict.count(s) != 0){
            encoded_str += haffman_dict[s];
        }
    }
    
    return encoded_str;
}
